File: PathoFroh/src/main/java/com/patho/main/action/handler/views/ReportView.java
package com.patho.main.action.handler.views;

import java.io.ByteArrayInputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import javax.faces.context.FacesContext;
import javax.faces.event.PhaseId;

import com.patho.main.config.PathoConfig;
import org.primefaces.model.DefaultStreamedContent;
import org.primefaces.model.StreamedContent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Configurable;

import com.patho.main.action.handler.GlobalEditViewHandler;
import com.patho.main.model.PDFContainer;
import com.patho.main.model.patient.DiagnosisRevision;
import com.patho.main.model.patient.Task;
import com.patho.main.model.util.audit.Audit;
import com.patho.main.repository.MediaRepository;
import com.patho.main.service.PDFService;
import com.patho.main.template.PrintDocument.DocumentType;
import com.patho.main.ui.task.DiagnosisReportUpdater;
import com.patho.main.util.helper.HistoUtil;
import com.patho.main.util.pdf.LazyPDFReturnHandler;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@Configurable
public class ReportView extends AbstractTaskView {

	@Autowired
	@Getter(AccessLevel.NONE)
	@Setter(AccessLevel.NONE)
	private MediaRepository mediaRepository;

	/**
	 * List of all diagnosis revisions with correspondending reports
	 */
	public List<DiagnosisReportData> data;

	/**
	 * The data which are displayed
	 */
	private DiagnosisReportData selectedData;

	/**
	 * The selected PDF container
	 */
	private PDFContainer selectedContainer;

	/**
	 * If true the update poll will be started
	 */
	private boolean generatingPDFs;

	/**
	 * If true the update poll will be stoped
	 */
	private boolean generationCompleted;

	/**
	 * True if selected pdf is currently generating
	 */
	private boolean generatingSelectedPDF;

	public ReportView(GlobalEditViewHandler globalEditViewHandler) {
		super(globalEditViewHandler);
	}

	public void loadView() {
		data = new ArrayList<DiagnosisReportData>();

		setGeneratingPDFs(false);
		setSelectedContainer(null);
		setGeneratingSelectedPDF(false);
		setGenerationCompleted(false);

		Task task = getTask();

		for (int i = 0; i < task.getDiagnosisRevisions().size(); i++) {
			DiagnosisRevision revision = HistoUtil.getNElement(task.getDiagnosisRevisions(), i);

			Optional<PDFContainer> c = PDFService.findDiagnosisReport(task, revision);

			if (!c.isPresent() && revision.getCompleted()) {
				logger.debug("No report present, generating new report");
				
				DiagnosisReportData reportData = new DiagnosisReportData(revision, true);
				task = new DiagnosisReportUpdater().updateDiagnosisReportNoneBlocking(task, revision, reportData);
				// serach for created pdfcontainer
				reportData.setPdf(PDFService.findDiagnosisReport(task, revision).get());
				data.add(reportData);
				setGeneratingPDFs(true);
			} else if (revision.getCompleted()) {
				data.add(new DiagnosisReportData(revision, c.get(), false));
			} else {
				PDFContainer container = new PDFContainer(DocumentType.DIAGNOSIS_REPORT_NOT_APPROVED,
						revision.getName(), PathoConfig.Companion.getREPORT_NOT_APPROVED_PDF(), PathoConfig.Companion.getREPORT_NOT_APPROVED_IMG());

				container.setAudit(new Audit());
				container.getAudit().setCreatedOn(revision.getAudit().getCreatedOn());

				data.add(new DiagnosisReportData(revision, container, false));
			}

		}

		if (data.size() > 0) {
			setSelectedPDF(data.get(0));
		}

		setTask(task);
	}

	public void updateData() {

		if (isGeneratingSelectedPDF()) {
			setSelectedPDF(getSelectedData());
		}

		boolean loading = getData().stream().anyMatch(p -> p.isLoading());
		
		setGenerationCompleted(!loading);
		setGeneratingPDFs(loading);
	}

	/**
	 * Returns the thumbnail als stream
	 * 
	 * @return
	 */
	public StreamedContent getThumbnail() {
		if (FacesContext.getCurrentInstance().getCurrentPhaseId() == PhaseId.RENDER_RESPONSE) {
			// So, we're rendering the HTML. Return a stub StreamedContent so that it will
			// generate right URL.
			return new DefaultStreamedContent();
		} else {
			// So, browser is requesting the image. Return a real StreamedContent with the
			// image bytes.
			String id = FacesContext.getCurrentInstance().getExternalContext().getRequestParameterMap().get("id");
			System.out.println("returning thumbnail");
			if (id != null && mediaRepository.isFile(id)) {
				byte[] img = mediaRepository.getBytes(id);
				return new DefaultStreamedContent(new ByteArrayInputStream(img), "image/png");
			} else
				return new DefaultStreamedContent(
						new ByteArrayInputStream(mediaRepository.getBytes(PathoConfig.Companion.getPDF_NOT_FOUND_PDF())), "image/png");
		}
	}

	public StreamedContent getPDF() {
		FacesContext context = FacesContext.getCurrentInstance();
		if (context.getCurrentPhaseId() == PhaseId.RENDER_RESPONSE || getSelectedContainer() == null) {
			// So, we're rendering the HTML. Return a stub StreamedContent so
			// that it will generate right URL.
			return new DefaultStreamedContent();
		} else {

			byte[] img;

			if (mediaRepository.isFile(getSelectedContainer().getPath()))
				img = mediaRepository.getBytes(getSelectedContainer().getPath());
			else
				img = mediaRepository.getBytes(PathoConfig.Companion.getPDF_NOT_FOUND_PDF());

			return new DefaultStreamedContent(new ByteArrayInputStream(img), "application/pdf",
					getSelectedContainer().getName());
		}
	}

	public void setSelectedPDF(DiagnosisReportData data) {
		logger.debug("Setting pdf " + data);
		if (data != null) {
			setSelectedData(data);
			if (data.isLoading()) {
				setGeneratingSelectedPDF(true);
				setSelectedContainer(null);
			} else {
				setGeneratingSelectedPDF(false);
				setSelectedContainer(data.getPdf());
			}
		}
	}

	@Getter
	@Setter
	@Configurable
	public static class DiagnosisReportData implements LazyPDFReturnHandler {

		protected final Logger logger = LoggerFactory.getLogger(this.getClass());

		@Autowired
		@Getter(AccessLevel.NONE)
		@Setter(AccessLevel.NONE)
		private MediaRepository mediaRepository;

		private DiagnosisRevision diagnosis;

		private PDFContainer pdf;

		/**
		 * True if the image is generated
		 */
		private boolean loading;

		public DiagnosisReportData(DiagnosisRevision diagnosis, PDFContainer pdf) {
			this(diagnosis, pdf, false);
		}

		public DiagnosisReportData(DiagnosisRevision diagnosis, boolean loading) {
			this(diagnosis, null, loading);
		}

		public DiagnosisReportData(DiagnosisRevision diagnosis, PDFContainer pdf, boolean loading) {
			super();
			this.diagnosis = diagnosis;
			this.pdf = pdf;
			this.loading = loading;
		}

		@Override
		public void returnPDFContent(PDFContainer container, String uuid) {
			if (loading) {
				loading = false;
			}
		}
	}
}

File: PathoFroh/src/main/kotlin/com/patho/main/action/views/ReportView.kt
package com.patho.main.action.views

import com.patho.main.config.PathoConfig
import com.patho.main.model.PDFContainer
import com.patho.main.model.patient.DiagnosisRevision
import com.patho.main.model.patient.Task
import com.patho.main.model.util.audit.Audit
import com.patho.main.service.PDFService
import com.patho.main.template.PrintDocument
import com.patho.main.template.PrintDocumentType
import com.patho.main.ui.task.DiagnosisReportUpdater
import com.patho.main.util.pdf.LazyPDFReturnHandler
import com.patho.main.util.pdf.StreamedContentInterface
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Component
@Scope(value = "session")
open class ReportView : AbstractTaskView(), StreamedContentInterface {

    /**
     * List of all reportIntent revisions with correspondending reports
     */
    open var data: MutableList<DiagnosisReportReturnHandler> = mutableListOf()

    /**
     * The data which are displayed
     */
    open var selectedData: DiagnosisReportReturnHandler? = null

    /**
     * The selected PDF container
     */
    override var selectedContainer: PDFContainer? = null

    /**
     * If true the update poll will be started
     */
    open var generatingPDFs: Boolean = false

    /**
     * If true the update poll will be stoped
     */
    open var generationCompleted: Boolean = false

    /**
     * True if selected pdf is currently generating
     */
    open var generatingSelectedPDF: Boolean = false

    /**
     * Returns a the thumbnail list for displaying the thumbnails
     */
    override val thumbnailList: List<PDFContainer>
        get() = data.map { p -> p.pdf ?: PDFContainer() }

    override fun loadView(task: Task) {
        super.loadView(task)

        logger.debug("Loading report view")

        data = mutableListOf()

        generatingPDFs = false
        selectedContainer = null
        generatingSelectedPDF = false
        generationCompleted = false

        for ((i, revision) in task.diagnosisRevisions.withIndex()) {

            val c = PDFService.findDiagnosisReport(task, revision)

            if (!c.isPresent && revision.completed) {
                logger.debug("No report present, generating new report")
                val returnHandler = DiagnosisReportReturnHandler(revision, true)
                // TODO updating task globally?
                this.task = DiagnosisReportUpdater().updateDiagnosisReportNoneBlocking(task, revision, returnHandler)
                // serach for created pdfcontainer
                returnHandler.pdf = PDFService.findDiagnosisReport(task, revision).get()
                data.add(returnHandler)
                generatingPDFs = true
            } else if (revision.completed) {
                data.add(DiagnosisReportReturnHandler(revision, c.get(), false))
            } else {
                val container = PDFContainer(PrintDocumentType.DIAGNOSIS_REPORT_NOT_APPROVED,
                        revision.name, PathoConfig.REPORT_NOT_APPROVED_PDF, PathoConfig.REPORT_NOT_APPROVED_IMG)

                container.audit = Audit()
                container.audit?.createdOn = revision.audit?.createdOn ?: System.currentTimeMillis()

                data.add(DiagnosisReportReturnHandler(revision, container, false))
            }

        }

        println(data +" ---- ")

        if (data.size > 0) {
            selectedData
            setSelectedPDF(data[0])
        }
    }

    /**
     * Is called via poll by the gui, checks if the selected pdf is generated or if it is ready
     */
    open fun updateData() {
        if (generatingSelectedPDF) {
            setSelectedPDF(selectedData)
        }

        val loading = data.any { p -> p.loading }

        generationCompleted = !loading
        generatingPDFs = loading
    }

    /**
     * Sets the selected PDF if it is available
     */
    open fun setSelectedPDF(data: DiagnosisReportReturnHandler?) {
        logger.debug("Setting pdf " + data!!)
        if (data != null) {
            selectedData = data
            if (data.loading) {
                generatingSelectedPDF = true
                selectedContainer = null
            } else {
                generatingSelectedPDF = false
                selectedContainer = data.pdf
            }
        }
    }

    /**
     * Return-Handler for pdf generation
     */
    open class DiagnosisReportReturnHandler : LazyPDFReturnHandler {

        open var diagnosis: DiagnosisRevision

        open var pdf: PDFContainer?

        /**
         * True if the image is generated
         */
        open var loading: Boolean = false

        constructor(diagnosis: DiagnosisRevision, pdf: PDFContainer?, loading: Boolean) {
            this.diagnosis = diagnosis
            this.pdf = pdf
            this.loading = loading
        }

        constructor(diagnosis: DiagnosisRevision, loading: Boolean) : this(diagnosis, null, loading) {}

        override fun returnPDFContent(container: PDFContainer, uuid: String) {
            if (loading) {
                loading = false
            }
        }
    }
}